src/wallet/test/coinselector_tests.cpp

```diff
@@ -997,7 +1005,10 @@ BOOST_AUTO_TEST_CASE(check_max_weight)
             chain, m_args);
 
         BOOST_CHECK(result);
-        BOOST_CHECK(has_coin(result->GetInputSet(), CAmount(50 * COIN)));
+        // Verify that only the 50 BTC UTXO was selected
+        const auto& selection_res = result->GetInputSet();
+        BOOST_CHECK(selection_res.size() == 1);
+        BOOST_CHECK(selection_res.begin()->GetEffectiveValue() == 50 * COIN);
```

Comment:
hmm that doesn't sound right.

As we set all feerates to 0, the change cost will be 0. So the waste calculation is simply a `waste = inputs total amount - target`.
So we actually will never select the SRD result, because SRD is forced to select the 50 btc UTXO to succeed (otherwise it surpasses the max allowed weight).

In other words, Knapsacks result always return a waste of 0.5 btc, while SRD will return a waste of 0.5 btc + every extra added input that appears on the shuffled vector prior reaching the 50 btc UTXO. So, knapsacks_waste <= srd_waste.

Reply:
No, the [waste](https://murch.one/posts/waste-metric/) is calculated via `waste = weight_of_inputs × (feerate - long_term_feerate_estimate)` where the LTFRE is 10 ṩ/vB. This means that at a feerate of 0 ṩ/vB, an input set with greater weight results in a lower (“more negative”) waste. I.e. if the SRD solution happens to roll an input set with 500 small inputs plus the 50 ₿ input it would fit in the standard transaction weight and have a better score than the input set with only the 50 ₿ input.

Reply:
> the [waste](https://murch.one/posts/waste-metric/) is calculated via waste = weight_of_inputs × (feerate - long_term_feerate_estimate) where the LTFRE is 10 ṩ/vB.

Where is the LTFRE set to 10ṩ/vB?

I'm seeing that it's set to 0:
https://github.com/bitcoin/bitcoin/blob/edc3d1b296e34838d649dc21b8483a52e214932a/src/wallet/test/coinselector_tests.cpp#L971-L972

So inside `GetSelectionWaste` the loop over all the inputs should be returning waste = 0 (waste += input_weight x 0). 
Then below, as `change_cost` is 0 too, we get into:
https://github.com/bitcoin/bitcoin/blob/edc3d1b296e34838d649dc21b8483a52e214932a/src/wallet/coinselection.cpp#L396-L400

Reply:
Ah, my bad. I did not expect anyone to change LTFRE in tests. :face_with_spiral_eyes:
In that case, the weight of the inputs is irrelevant for the waste score (because `weight × (0 - 0) = 0`), and it will only prefer a changeless input set over one with change. Then it depends on the order in which results are evaluated in `ChooseSelectionResult(…)`. I still don't think this is a good idea.

It would be better to make a test that only uses Knapsack to test the behavior of Knapsack is as expected, instead of depending on the order of evaluation in another function and setting global constants to odd values.

Reply:
> In that case, the weight of the inputs is irrelevant for the waste score (because weight × (0 - 0) = 0), and it will prefer an input set with change over one without change (unless excess is also 0). So very likely, the outcome of this test depends on either SRD being unsuccessful and/or that the order in which results are evaluated in ChooseSelectionResult(…) prefers the Knapsack result.

Actually, the result does not depend on the order in which the algorithms are executed. SRD will always succeed and will always have a higher waste score than snapsack (only could fail if after shuffling the vector, the big UTXO is located at last due the exceeded max weight).

The feerates are 0, same as the change cost, so the waste calculation is just a simple: 
```c++
waste = selected_effective_value - target;
```

So, Snapsack result only contains the 50 BTC UTXO (waste = 0,5 btc).
And, SRD result has the 50 BTC UTXO + all the other small UTXOs that appear in the shuffled vector prior to the big UTXO (waste = 0,5 btc + all smaller UTXOs).

Essentially, the best selection that SRD could make is when the big UTXO appears first after shuffling the coins vector, which is equal to the only solution that snapsack returns. All the other solutions that SRD could return have a higher waste than snapsack.

> Since we are explicitly testing the behavior of Knapsack, this should be a test that just runs Knapsack and checks that the Knapsack behavior is expected.

Make a specific test for this sounds good too. But I still think that would be nice to make tests that use the entire coin selection process and check that the expected result is returned (not only checking that any valid result is retrieved): check the inputs size, the value of them, the algorithm used etc.

> Having thought some more about this, setting LTFRE and cost_of_change to 0 basically turns off BnB. It will then only generate solutions that match the target to the satoshi. Definitely feels off that behavior of coin selection algorithms is so fundamentally changed in coinselector_tests.cpp. 🤔

So we have a new working path hehe (probably more suited for a follow-up PR): validate BnB further by reviewing all the test running the entire coin selection process with the feerates and change_cost in 0.
Plus, would be nice to code new tests running coin selection and expecting the best result to be retrieved by BnB.

Reply:
I think you may be referring to the description before BnB, which only describes the case for changeless.

If you check `GetSelectionWaste(…)`, you'll see that the formula is `waste = input_weight × (feerate - LTFRE) + excess + cost_of_change`. `excess` is the amount of sats dropped to the fees because we overshot the selection target but do not have enough for a change output. It can only be non-zero in changeless solutions. `cost_of_change` is set to zero in changeless solutions, since we do not create change. So, the score will only ever include a non-zero value for one of `excess` or `cost_of_change`. Since both SRD and Knapsack create change, both add `cost_of_change` which is zero here. Since `LTFRE == feerate`, the input weight also gets multiplied with 0. So any input set for both these algorithms will always have a waste score of 0.

Btw, I first described the waste metric.